core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java

```diff
@@ -23,15 +38,61 @@ public class ApplicationTypeLoader implements TypeLoader {
 
   private final TypeLoader primitivesTypeLoader = new PrimitiveTypesTypeLoader();
   private final TypeLoader specialTypesLoader = new SpecialTypesTypeLoader();
+  private final Map<String, String> extensionModelNamesByPrefix;
+  private final TypeCatalog dependenciesTypeCatalog;
+
+  public ApplicationTypeLoader(Set<ExtensionModel> extensionModels) {
+    this.dependenciesTypeCatalog = getDefault(extensionModels);
+
+    extensionModelNamesByPrefix = new HashMap<>(extensionModels.size());
+    for (ExtensionModel extensionModel : extensionModels) {
+      extensionModelNamesByPrefix.put(extensionModel.getXmlDslModel().getPrefix(), extensionModel.getName());
+    }
+  }
 
   @Override
   public Optional<MetadataType> load(String typeIdentifier) {
-    // TODO: MULE-20071 - implement app type catalog lookup.
     Optional<MetadataType> primitive = primitivesTypeLoader.load(typeIdentifier);
     if (primitive.isPresent()) {
       return primitive;
+    }
+
+    Optional<MetadataType> special = specialTypesLoader.load(typeIdentifier);
+    if (special.isPresent()) {
+      return special;
+    }
+
+    return getFromDependency(typeIdentifier);
+  }
+
+  // The string format can be the full name of the type, or a string with syntax <extension-prefix>:<type-alias> if
+  // the type has an alias. This format is temporal in order to use the types from the test operations, but may change
+  // when the type catalog using dataweave is fully implemented.
```

Comment:
Do we have a ticket to reference to that?

Reply:
Added the reference